design-proposals/klip-42-schema-migrations-tool.md

```diff
@@ -0,0 +1,442 @@
+| LATEST      | 2       | Add users     | Undone   | No       | <MD5-sum>  | 12-03-2020 10:34:30 | 12-03-2020 10:34:34 | 1        |
++-------------+---------+---------------+----------+----------+------------+---------------------+---------------------+----------+
+```
+
+The tool will later use pull queries to identify the current state of the system:
+```sql
+SELECT version, state, previous FROM SCHEMA_VERSION WHERE version_key = 'CURRENT'; 
+```
+
+When undo changes happen, the query will first get the current schema version to undo. Apply the revert operations, then set `CURRENT` to
+the previous migrated version. In this case, it will look at the `previous` column in `SCHEMA_VERSION`, then use a pull query to get information
+about the previous migration.
+```sql
+SELECT * FROM SCHEMA_VERSION WHERE version_key = '<previous>';
+```
+
+And set the new `CURRENT` in the `MIGRATION_EVENTS` to update the current schema version.
+
+### Undo migrations
+
+Undo a previous migration is necessary in the application lifecycle. A user sometimes want to revert the changes of an application because of a bug found in it.
+This also requires the database schema to be reverted or rollback to the previous version.
+
+The migrations tool allows users to revert changes. Note that undo is considered a forward migration. A new migration file is required which contains
+the SQL operations to revert the changes of a previous migration. The file in this case should contain the `U` prefix with the version number of the migration to rollback.
+
+For instance, let's undo the migration version 2 applied before. The migration file used before was named `V2__Add_users.sql`. For the undo file, we should add the `U` prefix.
+
+`U2_Add_users.sql`
+```sql
+DROP TABLE users;
+```
+
+The user runs the migration tool on the ksqlDB cluster. The tool detects the cluster has version 2, so it executes only the undo file for version 2.  
+It then sets the state of the cluster to version 1.
+
+```shell script
+$ ksql-migrations undo
+Current version of schema: 2
+Undoing migration of schema to version 2 - Add users
+```
+
+The `undo` action will only revert the previous change. It will not attempt to undo all applied migrations.
+
+
+### Naming convention
+
+Hope you have noticed the naming rules I followed in the previous examples for naming the migration files. Having a naming convention for files is necessary
+for the tool to detect what migration to apply or revert. Naming files are easier than creating configurations or command parameters to specify the same info.
+
+The migration files follow the same `Flyway` naming convention.
+`(Prefix)(Version)__(Description).sql`
+
+The `Prefix` specifies whether the file is a new migration (`V`) or undo (`U`) file.
+The `Version` is the version number used for the schema. For minor versions, such as `1.1`, an underscore is required (i.e. `1_1`)
+The `Description` is just a name or description of the new migration.
```

Comment:
Is there any ceremony around the description, or can it be whatever? It looks like that is how you make those nice printouts when you migrate?

Reply:
Just using underscores or spaces to separate words. If underscores are used, then they will automatically converted to spaces at runtime. I'll add that comment.